File: nodejs/day04/index1.js
const fs = require("fs");

const lines = fs
  .readFileSync("./input", "utf-8")
  .trim()
  .split("\n");

const isValid = line => {
  const hash = {};
  const words = line.split(" ");
  for (const word of words) {
    if (hash[word]) {
      return false;
    }
    hash[word] = true;
  }
  return true;
};

console.log(lines.filter(isValid).length);

File: clojure/day04/index1.clj
(require '[clojure.string :as s])

(def parse-line #(s/split % #" "))

(def lines (->> "./input"
  slurp
  s/trim
  s/split-lines
  (map parse-line)
))

(defn is-valid* [acc words] (
  if (== (count words) 0) true (
    let [
      w (first words)
      words' (rest words)
    ](if (get acc w) false 
      (is-valid* (assoc acc w true) words')
    )
  )
))

(defn is-valid [words] (is-valid* {} words))

(println (->> lines
  (filter is-valid)
  count
))
